### DISCUSSION TITLE
Dynamic mod multipliers

### DISCUSSION THREAD

**Discussion by Nyanaro** [ppy/osu#25706] [Ideas]
I am making the following post to propose dynamic mod multipliers for osu!lazer, meaning that mod multipliers for select mods would be based on the map that is being played, rather than a static number.

First, here are some of the reasons why I feel dynamic multipliers are better than static ones;

- The difficulty of mods differs greatly per map. For example a map with lower approach rate is generally more difficult with hidden than one with high approach rate, as the notes are hidden for longer. In the case of hard rock, circle size is increased based on the circle size difficulty setting value, rather than the radius of the hitcircle. This creates inconsistencies such as a map with CS3 + HR having its circles' radii decreased by around 1,11x, but with CS7 + HR, this increases to a massive 1,69x.
- For some beatmaps, mods do not change anything. Maps such as [Ryuusei Aika's LUVORATORRRRRY](https://osu.ppy.sh/beatmapsets/1296266#osu/2689496) Hard Rock only flips the map upside down, it doesn't change the difficulty at all, similarly EZ only gives the player 3 lives.
- For some maps, some mods are very overpowered, and "ruin" leaderboards, a great example of this is [Apo11o's set of Agony](https://osu.ppy.sh/beatmapsets/1697454#osu/3468026), where EZ increases the radius of the circles by 3,33x in the top difficulty, making the map trivial compared to nomod.
- On high approach rates, especially with doubletime on AR11, hidden is considered by most to be a "preference mod" and it having a static multiplier makes highscores on certain leaderboards require it to be enabled in order to compete.

I believe what I propose below solves all of these issues, or at the very least alleviates them greatly.

I have created formulas for dynamic multipliers for Hard Rock, Hidden, Double Time, and Easy. I have been receiving feedback from Bubbleman's Lazer Feedback discord for the formulas for the past 3 months and I feel ready to release it here now. The formulas are each meant to be simple, and easy to understand, only being based on a map's difficulty settings and nothing else.

The constants used are based off of common freemod mappooling practices, where for example "CS5 AR8" maps are considered to be equal in difficulty for both Hidden and Hard Rock, as well as general community opinion, which was based upon feedback gathered in the aforementioned discord. Further explanation [here](https://github.com/ppy/osu/discussions/25706#discussioncomment-7844669)

Below are the formulas for each of the 4 mods.

Hidden
The formula below sets the Hidden mod multiplier based on the map's AR difficulty setting. It factors in the amount of time an object spends faded out. The formula uses logarithms to set an "upper bound" for extremely low approach rates, to account for memorization, density, and how players at approach rates this low switch from reacting to objects appearing to reacting to objects fading out. Refer to the spreadsheet linked below to see the results of how the values change depending on a map's difficulty settings. This results in a higher hidden multiplier for low approach rates, and a lower hidden multiplier for high approach rates.
![HD](https://github.com/ppy/osu/assets/22597387/fc626292-d467-48b5-8e06-cfe791756ce5)

[Graph (Desmos)](https://www.desmos.com/calculator/exag8uz0mu) (credit to Bubbleman)

FO = Time object spends faded out
HDC = "Hidden Control" (set to AR8)
HD multiplier is applied after other mods (in the case of DT, and HR)

Hard Rock
The formula below sets the Hard Rock mod multiplier based on the map's OD and CS difficulty settings. The OD and CS components of the formula are calculated separately, and then multiplied with each other at the end. The CS component factors in the change of a circle's radius, and the OD component factors in the change in hit window. Refer to the spreadsheet linked below to see the results of how the values change depending on a map's difficulty settings. This results in a lower hard rock multiplier when the CS and OD don't change much, and a higher multiplier when they do.

[Graph (Desmos)](https://www.desmos.com/calculator/xzvpk7akrv) (credit to Bubbleman)

![HR](https://cdn.discordapp.com/attachments/148120887751606272/1184171902617530369/image.png)

RI = radius increase
HWR = hit window reduction
HRCSC = "Hard Rock Circle Size Control" (set to CS5)					
HRODC = "Hard Rock Overall Difficulty Control" (set to OD8)					

Easy
The Easy multiplier works the same as the Hard Rock multiplier, only in inverse. Refer to the spreadsheet linked below to see the results of how the values change depending on a map's difficulty settings.

[Graph (Desmos)](https://www.desmos.com/calculator/5hd3mnmenj) (credit to Bubbleman)

![EZ](https://github.com/ppy/osu/assets/22597387/9c98638d-ca26-4cfd-91b3-a1b93b6e66ae)

RD = radius decrease
HWI = hit window increase
HRCSC = "Hard Rock Circle Size Control" (set to CS5)					
HRODC = "Hard Rock Overall Difficulty Control" (set to OD8)					

Double Time
The Double Time multiplier takes into account the rate changes used, as well as the approach rate of the map. There is an added "reaction bonus" for extremely high approach rates, starting at AR9+DT. Refer to the spreadsheet linked below to see the results of how the values change depending on a map's difficulty settings.

![DT](https://github.com/ppy/osu/assets/22597387/37b8c935-534e-4a81-9a59-24705837fdf0)

ApproachTime = Approach rate in milliseconds after DT is applied					

**I have also made a spreadsheet containing various tables and info showing what the multipliers are for any given difficulty settings. You can find it [here](https://docs.google.com/spreadsheets/d/1fxTxPd0kL67rrZz-CB4mc6l7hE8t3cpvfERfa750_co/edit#gid=1152887626).** I do however recommend you to create your own methods of showing the formulas (I'm not the best at making easy-to-digest spreadsheets). If you also have better ideas for formulas, or want to adjust the values of the formulas, I encourage you to do so.

Additionally, in the spreadsheet I have a for the rate multipliers for Half Time.

Below I have some examples of sets of difficulty settings beatmaps and what their multipliers for some mod combinations would be.

The sheet also contains a page called "other multipliers" which has work-in-progress proposals for mod multipliers and rankability of new lazer mods, as well as multipliers for existing mods not included in the dynamic mod multipliers proposal. The page is currently hidden, as it isn't relevant for this particular discussion.

With this proposal, I hope to alleviate leaderboard issues caused by static multipliers. Thanks for reading.

:D

**Comment by SimredReal** [comment]
yeah

**Comment by bdach** [comment]
This feels like inventing another performance system but for score. I do not think that should be a thing.

**Reply by Bubbleman532** [reply]
The idea behind this is that these mod calculations are entirely based on physical traits of the game which in theory should never change - having an unchanging multiplier when mods such as HR affect a map on a linear scale (with a cutoff) correlating to a map stat doesn't treat each case fairly

**Reply by bdach** [reply]
Would this be alleviated by a performance points leaderboard instead?

**Reply by Bubbleman532** [reply]
It would not, since performance point rating doesn't currently correctly take into account density and likely never will since the rating is geared to mechanical performance rather than comprehension

**Reply by Nyanaro** [reply]
A performance point leaderboard isn't what is attempted here, albeit it would achieve similar results (to an extent).

Most players expect a score leaderboard to be fairly static and not affected by changes to the performance point system. If a performance points leaderboard were to be added, the order of scores on leaderboards would change each update. This system is meant to be a more permanent fixture, like score leaderboards are currently.

Performance points are also meant to be compared with each other outside the context of an individual map. Scoring is confined within the map leaderboard. They achieve very different purposes.

**Reply by TextAdventurer12** [reply]
> take into account density and likely never will

this is false, there is a prototype rework which aims to tackle density and general reading difficulty. Ideally, PP is a measure of difficulty to set a play, not just 'mechanical performance'

**Reply by Rekunan** [reply]
> Performance points are also meant to be compared with each other outside the context of an individual map

While this is true, it doesn't mean you can't compare pp within a map leaderboard. Especially considering all the new possible mod combinations, it will be way easier to compare using pp instead of score (ex: DT 1.3x vs DA cs6).

**Reply by bad-faith** [reply]
Performance point leaderboard would suck, because DT passes would go on top on NM full comboes and what not. As PP improves passing a map on DT will inevitably give more PP than an HDHR SS.

What we need is a score multiplier based on star rating change.

**Comment by TextAdventurer12** [comment]
The best part of score in my opinion is that it's very simple for new players to understand. This removes most all of that simplicity, and still lacks the analysis of pp. So you end up with a system which is a worse approximation for skill required than pp, and harder to understand than score.

**Reply by Nyanaro** [reply]
While this is a more complex system than simply having static multipliers, I believe most players will develop an intuitive understanding for these multipliers fairly quickly, as they only depend on a map's difficulty settings and nothing more.

It looks like complicated math jargon on the surface, but in reality all it does is "the more HR changes circle size, the bigger the multiplier"

**Reply by rushiiMachine** [reply]
> very simple for new players to understand

Personally I don't think this is really the case in the first place, most new players don't even notice that mods even give a score multiplier assuming they even know mods exist. Using new players as a reference frame here is really irrelevant since most don't even play mods at all, and if they do the specifics really don't matter since they're never making it onto a leaderboard.

**Reply by TextAdventurer12** [reply]
> new players don't even notice that mods even give a score multiplier

There is a very obvious box that says what the current score multiplier is in mod select. I'm not sure how anybody is going to miss that

**Comment by peppy** [comment]
won't happen under my watch

**Reply by Bubbleman532** [reply]
Can I ask why

**Reply by isakvik** [reply]
I think you're either overcomplicating or misunderstanding what this proposal is about if you're so dismissive about it. Fundamentally it's the exact same process of having a different score multiplier for DT rateups such as 1.4x vs 1.5x; faster speed should give you a bigger score because **the mod has a bigger effect on gameplay**, and this addresses that for difficulty settings. It shouldn't be antithetical to goals for score at all, though if it is, please explain why.

**Reply by bdach** [reply]
i'm sorry but in what world is this "overcomplicating" given that the opening post is literally plastered with formulae?

**Reply by bdach** [reply]
okay i will attempt to qualify the above - which can probably be interpreted as inflammatory - as my _personal_ objection to this idea:

this makes the mod system inscrutable, bringing it to the level of pp inscrutability. it is not friendly for anybody except pro players. the game has pp for pro players. we've already yielded on simple score for your sake, please do not push everything else further into being inscrutable.

like there are so many formulae in the OP without even a sliver of a qualification of why those in particular. where did any of this come from? the declaration that these multipliers are going to be set in stone or "more permanent" also - i'm sorry - reeks of naivete, as i guarantee you someone is going to find some map where this is broken in some respect and something is going to have to be done about it, because attempting to approximate the difficulty of a map based on mods _is already a task that the pp algorithms attempt - and sometimes fail - to do correctly_!

**Reply by Bubbleman532** [reply]
As a rebuttal:

- In the case of hard rock, circle size is increased based on the circle size difficulty setting value, rather than the radius of the hitcircle. This creates inconsistencies such as a map with CS3 + HR having its circles' radii decreased by around 1,11x, but with CS7 + HR, this increases to a massive 1,69x.

This difficulty change is not approximate - **none of this is based on anything arbitrary**, unless you want to try and argue that small circles are just as easy to hit as big circles. If you want to change mods to match static mod multipliers, be my guest, however the way that HR specifically works now is actually exponential in nature rather than linear, and also has a limit it can hit.

Other calculations are built on other mod stats and map stats, and again are not arbitrary. This is not for comparison between maps, this is in correlation to the direct effect on the same map the mod has.

**Reply by bdach** [reply]
the explanation is appreciated, although it doesn't even cover the entirety of the HR stuff (there's also some sort of carve-out there for hitwindows, too). something approaching the above _human-understandable_ description of what is being done would definitely have done this proposal much more good than the wall of formulae in the OP.

if i am presented with formulae without explanation i will without fail call them arbitrary. provide the work, start to end.

that is mostly an aside, though; the inscrutability point still stands. as broken or bad flat multipliers may be in the general case, they are flat multipliers. they do not require a wiki page for explanation. this is helpful for accessibility and understanding. it is math that anybody can do with any old calculator. there are tradeoffs to be made here and the game does not and _should not_ aspire to cater only to the highest echelon. as long as the multipliers are good enough across the board flat - as good as they can be in such form - i do not see the point in attempting to reproduce the performance points algorithms - which again, _should be accounting for things such as circle size_ - with a secondary multiplier system.

**Reply by Nyanaro** [reply]
There seems to be some confusion as to what the formulae do, so I added short paragraphs explaining each of them.
https://github.com/ppy/osu/discussions/25706#discussioncomment-7832092

**Reply by Newbie1st** [reply]
What about implementing that for mods that specifically remove elements of the gameplay, like Spun Out, Constant Speed, No Release and Classic? I'd argue that they shouldn't affect score if they don't change the map, like if the map has no spinners, SV changes or sliders. To add to that, keeping the Classic mod multiplier the same for all maps seems like a bad idea, so what about making it scale depending on the slider/LN percentage instead?

**Comment by Nyanaro** [comment]
I added a short paragraph under each of the multipliers, explaining what each of them do. If you have any questions about specifics, please don't hesitate to ask.

I encourage you to visit the linked spreadsheet to see what kind of results the formulae produce, given a map's difficulty settings.

**Reply by Nyanaro** [reply]
The constants found in the formulae are of course up for debate, in addition to the formulae themselves of course. If you can find a more elegant and simple solution that produces better results, I'd be happy to see them.

While I have tried to balance them to the best of my ability, I am by no means attached to any of them. Please do offer suggestions on how to improve them. You can also make a copy of the spreadsheet and adjust the values as you see fit to experiment. If you have trouble navigating the spreadsheet when trying to do so, feel free to create your own, or send me a message.

**Reply by Nyanaro** [reply]
When editing the Hard Rock formula before posting it here, I made a mistake and forgot to add a constant, sorry, that must have been confusing. Fixed now

edit: clarified hidden formula image, also spotted a mistake there. my error.
edit: bubbleman noticed an error in the ez formula, fixed that

**Reply by bdach** [reply]
> The constants found in the formulae are of course up for debate, in addition to the formulae themselves of course

well, i mean, they are the _primary_ thing up for debate here. i appreciate the attempt to expound on the raw formulae but what you have right now doesn't help because it is mostly a dry restating of the formulae.

i can read math notation, but the math notation doesn't tell me *why* constant X has been chosen. or why there is a log in formula Y. or why factors foo and bar have been multiplied together. i don't know if there is any actual intuition behind these formulas, or if it's just a "looked good to me lol" type deal.

and if it _is the second_, then that majorly damages my confidence in the entire thing, because due to being heuristically chosen the formulae are not going to be as permanent as you promise, because there is _no_ objective reason not to tweak constant Z to whatever other value, because of some map quux not behaving correctly.

**Reply by Nyanaro** [reply]
The constants were all based on where the "playerbase" collectively agrees (as seen in tournament mappooling habits for the freemod slot) current mod multipliers being justified. For instance, the constants used aim to bring CS5 AR8 OD8 to a similar balance currently seen in ScoreV2, as those are the most agreed upon difficulty settings where both Hidden, and Hard Rock apply. Similarly, the EZ multiplier tries its best to correspond to mappooling habits of custom multipliers of tournaments (where tournaments retroactively multiply scores for certain mods (usually EZ), depending on the map in question. The DoubleTime reaction multiplier was based on a poll I conducted in Bubbleman's Lazer feedback discord.

I added a short paragraph, and linked this comment to the original post.

**Reply by bdach** [reply]
so to sum up, the constants are a byproduct of:

- mappooler consensus (not necessarily reflective of the general player base)
- poll with unknown number of respondents (guessing low, judging from the venue where it was conducted), likely selection biased towards the high end of the player base (for the same reason)

i don't know. doesn't feel very objective to me.

> custom multipliers of tournaments

another tangential topic here is that we have not discounted having mod multipliers be customisable in specific multiplier lobbies in the past. this could serve as a potential out that allows the competitive community to use its formulae as it sees fit without forcing the system on the entirety of the playerbase.

**Reply by Nyanaro** [reply]
> i don't know. doesn't feel very objective to me.

It's far more "objective" than anything else that has been proposed or anything that is currently implemented.

**Comment by DarkbitNetwork** [comment]
I think the formula used in this proposal is way too complicated or poorly explained, even I have a hard time understanding the meaning, and purpose of a variable and how you calculate it. For example, what IS "Time object spends faded out" exactly? Does it mean 100% transparency? and how do you calculate it? Also, many of the constants (that aren't "control" constant) are just there for some reason that I can't figure out. I was trying to map the formula to Desmos but gave up because it's so hard.

Anyway, aside from this, there are many reasons why I cannot accept this proposal. Though some people might have different reasons or opinions:
1. I believe the mod multiplier should only be based on its configuration (_cough cough DT/HT cough_) and the ruleset. Seeing the multiplier being different on every beatmap makes it very hard or nearly impossible to figure out the exact multiplier without knowing what beatmap it is.
1.1. I don't believe dynamic mod multiplier is the same as configuring a mod to have a different multiplier (like DT with 1.4x rate, etc.). Could you imagine trying to explain why a mod multiplier on this beatmap is different than having the same mod on another beatmap? 
With a static multiplier, you can at least explain that "dt multiplier is increased based on x speed" whereas, with dynamic mod multiplier, you have to explain something like "dt multiplier is increased depending on the beatmap's ar after dt and x speed" and that's with the most simple formula of all.
2. The difficulty of mods on certain beatmaps (proposal's reason 1, 2, 3) being different by either easier or harder on certain beatmaps when compared to using the same mod on other beatmaps. I think this should be the job of the pp system rather than tweaking the multiplier until everything is right which is what pp system should be doing. Making the mod multiplier different on every beatmap would probably run into the same problem that pp system has and probably defeat the whole point of dynamic mod multiplier.
3. The "requiring HD mod to be enabled to compete" (proposal's reason 4) can be fixed to certain extents without having to use a dynamic mod multiplier by simply changing the static mod multiplier so that using other mods combinations without HD (like HRDT, DTFL, etc.) would give more score than with mod combination with HD (like HDHR, HDDT, etc.).
4. As for certain mods "ruining" beatmap's leaderboard (proposal's reason 3), I think this is more of an issue with the scoring than the mods. I think this issue wouldn't even be present if you use score v2 or the current lazer master. (post #24166)

**Reply by Nyanaro** [reply]
> For example, what IS "Time object spends faded out" exactly? Does it mean 100% transparency? and how do you calculate it?

Yes it means 100% transparency and it is simply 0,3x the approach time.

> I was trying to map the formula to Desmos but gave up because it's so hard.

I'm in the process of doing that with Bubbleman, I'm somewhat busy with exams at the moment so I haven't finished all of them but [here is Hard Rock](https://www.desmos.com/calculator/jjgzcfn10s). I'll edit the post once they're all done.

> Could you imagine trying to explain why a mod multiplier on this beatmap is different than having the same mod on another beatmap?

It would be very easy to explain why CS0+HR has a smaller multiplier than CS4+HR, because CS0+HR changes nothing, and CS4+HR changes a lot.

> The "requiring HD mod to be enabled to compete" (proposal's reason 4) can be fixed to certain extents without having to use a dynamic mod multiplier by simply changing the static mod multiplier so that using other mods combinations without HD (like HRDT, DTFL, etc.) would give more score than with mod combination with HD (like HDHR, HDDT, etc.).

This is a reasonable alternative, and while I think it's effectively the same thing, and just as "confusing", it would also work to alleviate some of the issues caused by multipliers, albeit less effectively.

**Reply by TextAdventurer12** [reply]
> effectively the same thing, and just as "confusing"

This would just be a matter of tweaking the constant multipliers of each mod, and would be no more confusing than the current system, and much less confusing than this proposal

**Comment by ArmaniDilbo** [comment]
I believe that having a dynamic mod multiplier in the case of leaderboards wouldn't be particularly helpful, especially with leaderboards sorted by pp already becoming a thing.  Rather, I think using this for tournaments would be much more productive as score is still used as the main metric there, and would change the way teams handled overmoding in tournaments, making for a much more dynamic tournament and tournament viewing experience.  In addition, this would avoid new players being confused by the new scoring system.

**Reply by rodrigopina360** [reply]
Yeah, it would most likely lead to the same confusion when new players are watching tournaments and don't understand why in a CS3 map HR is worth less than HD, but at skill level of players playing a tournament, most would already know why, even if they don't know the exact formula and can't calculate exactly why a multiplier is that specific number.
Overall I also see the issue that people are currently already used to how multipliers work, making it so someone would need to check what would the HD, HR, DT or EZ multipliers be individually, which I think even for tournament play to be a horrible way to go around it. Maybe showing what the multipliers would be directly on each of the mods on the mod select screen to fix that, but then it would make the whole UI more complicated to digest.

**Comment by Nyanaro** [comment]
Added Desmos graphs to HR, HD, and EZ multipliers. DT WIP
Thank you Bubbleman for the graphs
